File: formacOS/Sources/AnyClipDaemon/PidLock.swift
import Foundation
import AnyClipCore

/// Single-instance lock shared with the Python implementation
/// (~/.anyclip/anyclip.pid, "<pid> <port>\n"). Port of
/// anyclip.prepare_pid_lock / release_pid_lock and helpers.
public enum PidLock {
    public static func prepare(port: UInt16, dir: URL) throws {
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let pidFile = dir.appendingPathComponent("anyclip.pid")

        // 1) PID file from a previous run (Python or Swift).
        if let content = try? String(contentsOf: pidFile, encoding: .utf8),
           let first = content.split(separator: " ").first,
           let oldPid = Int32(first.trimmingCharacters(in: .whitespacesAndNewlines)),
           oldPid > 0, oldPid != getpid(), processAlive(oldPid) {
            AnyLog.shared.info("another anyclip detected (pid \(oldPid) via PID file); terminating")
            guard terminate(oldPid) else {
                throw FatalStartupError(
                    "could not terminate previous anyclip (pid \(oldPid)); "
                    + "please run: kill -9 \(oldPid)")
            }
            AnyLog.shared.info("previous anyclip (pid \(oldPid)) terminated")
        }

        // 2) Stale state: port held without a matching PID file.
        if let listenerPid = findListeningPid(port: port), listenerPid != getpid() {
            if isAnyclipPid(listenerPid) {
                AnyLog.shared.info("anyclip listening on tcp/\(port) (pid \(listenerPid)); terminating")
                guard terminate(listenerPid) else {
                    throw FatalStartupError(
                        "could not terminate anyclip on tcp/\(port) (pid \(listenerPid)); "
                        + "please run: kill -9 \(listenerPid)")
                }
                usleep(300_000) // let the OS release the socket
            } else {
                throw FatalStartupError(
                    "tcp/\(port) is held by a non-anyclip process (pid \(listenerPid)); "
                    + "stop that process or quit it first")
            }
        }

        // 3) Record our pid (and chosen port for diagnostics).
        try? "\(getpid()) \(port)\n".write(to: pidFile, atomically: true, encoding: .utf8)
    }

    /// Remove our PID file, but only if it still points at us.
    public static func release(dir: URL) {
        let pidFile = dir.appendingPathComponent("anyclip.pid")
        guard let content = try? String(contentsOf: pidFile, encoding: .utf8),
              let first = content.split(separator: " ").first,
              Int32(first.trimmingCharacters(in: .whitespacesAndNewlines)) == getpid()
        else { return }
        try? FileManager.default.removeItem(at: pidFile)
    }

    static func processAlive(_ pid: Int32) -> Bool {
        guard pid > 0 else { return false }
        if kill(pid, 0) == 0 { return true }
        return errno == EPERM // exists, owned by another user
    }

    /// Pure matcher, exposed for tests. Case-insensitive so it recognises
    /// both `anyclip.py` and `AnyClip.app` command lines.
    static func argsLookLikeAnyclip(_ args: String) -> Bool {
        args.lowercased().contains("anyclip")
    }

    static func isAnyclipPid(_ pid: Int32) -> Bool {
        guard let out = runCommand("/bin/ps", ["-p", "\(pid)", "-o", "args="]) else {
            return false
        }
        return argsLookLikeAnyclip(out)
    }

    static func findListeningPid(port: UInt16) -> Int32? {
        guard let out = runCommand(
            "/usr/sbin/lsof", ["-nP", "-iTCP:\(port)", "-sTCP:LISTEN", "-t"])
        else { return nil }
        for line in out.split(separator: "\n") {
            if let pid = Int32(line.trimmingCharacters(in: .whitespaces)) { return pid }
        }
        return nil
    }

    /// SIGTERM, wait up to 2 s, then SIGKILL. True if the pid is gone.
    static func terminate(_ pid: Int32) -> Bool {
        if kill(pid, SIGTERM) != 0, !processAlive(pid) { return true }
        for _ in 0..<20 {
            usleep(100_000)
            if !processAlive(pid) { return true }
        }
        kill(pid, SIGKILL)
        for _ in 0..<10 {
            usleep(100_000)
            if !processAlive(pid) { return true }
        }
        return !processAlive(pid)
    }

    private static func runCommand(_ path: String, _ args: [String]) -> String? {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: path)
        process.arguments = args
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = FileHandle.nullDevice
        do { try process.run() } catch { return nil }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        return String(data: data, encoding: .utf8)
    }
}

File: forwindows/src/AnyClipApp/PidLock.cs
using System.Diagnostics;
using AnyClip.Core;

namespace AnyClip.App;

/// Shared ~/.anyclip/anyclip.pid ("<pid> <port>\n"). Windows semantics
/// follow anyclip.py: the pid-file evidence is trusted (_is_anyclip_pid
/// returns True on win32; no lsof port probe). Kill, wait 2 s, 0.3 s
/// socket settle.
public sealed class WindowsPidLock(string? dir = null) : IPidLock
{
    private readonly string _dir = dir ?? ConfigStore.DefaultDir();
    private string PidFile => Path.Combine(_dir, "anyclip.pid");

    public void Prepare(int port)
    {
        Directory.CreateDirectory(_dir);
        if (File.Exists(PidFile))
        {
            var first = File.ReadAllText(PidFile).Split(' ').FirstOrDefault();
            if (int.TryParse(first, out int oldPid)
                && oldPid > 0 && oldPid != Environment.ProcessId
                && TryGetProcess(oldPid) is { } proc)
            {
                RotatingLog.Shared.Info(
                    $"another anyclip detected (pid {oldPid} via PID file); terminating");
                try
                {
                    proc.Kill();
                    if (!proc.WaitForExit(2000))
                        throw new FatalStartupException(
                            $"could not terminate previous anyclip (pid {oldPid})");
                    RotatingLog.Shared.Info($"previous anyclip (pid {oldPid}) terminated");
                    Thread.Sleep(300); // let the OS release the listen socket
                }
                catch (FatalStartupException) { throw; }
                catch (Exception e)
                {
                    RotatingLog.Shared.Warning($"terminate pid {oldPid} failed: {e.Message}");
                    // Parity with the Python/Swift ports (FatalStartupError):
                    // if the old daemon is still alive we must not start a
                    // second instance against the same port/state dir.
                    using var still = TryGetProcess(oldPid);
                    if (still is not null)
                        throw new FatalStartupException(
                            $"could not terminate previous anyclip (pid {oldPid})");
                }
                finally { proc.Dispose(); }
            }
        }
        try { File.WriteAllText(PidFile, $"{Environment.ProcessId} {port}\n"); }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        { RotatingLog.Shared.Warning($"could not write PID file {PidFile}: {e.Message}"); }
    }

    public void Release()
    {
        try
        {
            if (!File.Exists(PidFile)) return;
            var first = File.ReadAllText(PidFile).Split(' ').FirstOrDefault();
            if (int.TryParse(first, out int pid) && pid == Environment.ProcessId)
                File.Delete(PidFile);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException) { }
    }

    private static Process? TryGetProcess(int pid)
    {
        try { return Process.GetProcessById(pid); }
        catch (ArgumentException) { return null; } // no such process
    }
}
